src/NuGet.Clients/NuGet.Tools/NuGetPackage.cs

```diff
@@ -1160,14 +1160,18 @@ private void BeforeQueryStatusForAddPackageForSolutionDialog(object sender, Even
                 {
                     isConsoleBusy = ConsoleStatus.Value.IsBusy;
                 }
-                // Enable the 'Manage NuGet Packages For Solution' dialog menu
-                // - if the console is NOT busy executing a command, AND
-                // - if the solution exists and not debugging and not building AND
-                // - if there are NuGetProjects. This means there are loaded, supported projects.
-                command.Enabled =
-                    IsSolutionExistsAndNotDebuggingAndNotBuilding() &&
-                    !isConsoleBusy &&
-                    await SolutionManager.Value.DoesNuGetSupportsAnyProjectAsync();
+
+                if (SolutionManager != null)
```

Comment:
Why is the solution manager null? 

If you look at the implementation of the method in question, it actually doesn't use the solution manager logic, it's really a method that could be converted to a static method if we really wanted to.

Reply:
> Why is the solution manager null?
> 

Problem very is similar to [PR#3889](https://github.com/NuGet/NuGet.Client/pull/3889) which I'm trying to fix timing bug, async init method `InitializeMEFAsync` still running but it already set[ `initialized = true` ](https://github.com/NuGet/NuGet.Client/blob/1729f1616e9b1812012bd4e0170ddffe1a2991ae/src/NuGet.Clients/NuGet.Tools/NuGetPackage.cs#L199), so `ShouldMEFBeInitialized` return false.

https://github.com/NuGet/NuGet.Client/blob/1729f1616e9b1812012bd4e0170ddffe1a2991ae/src/NuGet.Clients/NuGet.Tools/NuGetPackage.cs#L1149-L1152

Since MEF injection not complete yet for `IVsSolutionManager` type `SolutionManager` still null.
I only experience this during debugging. I can make same fix as [PR#3889](https://github.com/NuGet/NuGet.Client/pull/3889) but currently this check null thing is good enough.

> If you look at the implementation of the method in question, it actually doesn't use the solution manager logic, it's really a method that could be converted to a static method if we really wanted to.

Are you referring below method? We can't convert to static method because [_asyncServiceProvider](https://github.com/NuGet/NuGet.Client/blob/1729f1616e9b1812012bd4e0170ddffe1a2991ae/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs#L383) need to be injected at [constructor](https://github.com/NuGet/NuGet.Client/blob/1729f1616e9b1812012bd4e0170ddffe1a2991ae/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs#L161) first from another MEF injection. It gets ugly if we go that direction.

https://github.com/NuGet/NuGet.Client/blob/1729f1616e9b1812012bd4e0170ddffe1a2991ae/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs#L374-L383

Reply:
There are ways to force MEF to satisfy the imports. 
We shouldn't be running into that race condition.

> Are you referring below method? We can't convert to static method because _asyncServiceProvider need to be injected at constructor first from another MEF injection. It gets ugly if we go that direction.


The NuGetPackage class itself is an async service provider.

Reply:
> at

Can you refer any examples how to force MEF to satisfy the imports? I prefer this one because it's minimal change and  doesn't change anything else.

Reply:
The MEF initialization is actually supposed to happen: https://github.com/NuGet/NuGet.Client/blob/5ce9c44823a8930dd0bf57adbf5c32914d8e4ae3/src/NuGet.Clients/NuGet.Tools/NuGetPackage.cs#L1149-L1152. 

> Since MEF injection not complete yet for IVsSolutionManager type SolutionManager still null.

I'm not confident your understand of the problem is correct. As such I don't think this is the best course of action at this point. 

Note that if SolutionManager is null, then the other problem you linked never comes into play since there's no object to be initialized.

Reply:
I wrapped MEF initializer with Reentrant lock and tested on my local it looks good, error disappears. After I remove it error reemerge.